soccer/gameplay/skills/goalside_mark.py

```diff
@@ -0,0 +1,194 @@
+    @property
+    def ratio(self):
+        return self._ratio
+
+    @ratio.setter
+    def ratio(self, value):
+        self._ratio = min(max(value, 0.0), 1.0)
+
+    #Max distance for inclusion of ratio in target point
+    #If robot distance exceeds this then it will target the closest point on the mark line
+    @property
+    def mark_line_thresh(self):
+        return self._mark_line_thresh
+
+    @mark_line_thresh.setter
+    def mark_line_thresh(self, value):
+        self._mark_line_thresh = value
+
+    # Overrides mark_robot with a static point
+    @property
+    def mark_point(self) -> robocup.Point:
+        return self._mark_point
+
+    @mark_point.setter
+    def mark_point(self, value: robocup.Point):
+        self._mark_point = value
+
+    #Robot to mark (what it sounds like)
+    @property
+    def mark_robot(self) -> robocup.Robot:
+        return self._mark_robot
+
+    @mark_robot.setter
+    def mark_robot(self, value: robocup.Robot):
+        self._mark_robot = value
+
+    # Choose a robot close to the mark point
+    def role_requirements(self):
+        req = super().role_requirements()
+        if self._target_point is not None:
+            req.destination_shape = self._target_point
+        return req
+
+    def get_mark_line(self):
+        #Finds the line segment between the ball and the highest danger shot point
+        #Cuts off the portion of the line that is inside of the goal box
+        #@return: LineSegment to defend on, shot_point
+
+        #Define the segments where the defender can go closest the goal
+        offset = constants.Robot.Radius
+        left_seg = robocup.Segment(
+            robocup.Point(-constants.Field.PenaltyLongDist / 2 - offset, 0),
+            robocup.Point(-constants.Field.PenaltyLongDist / 2 - offset,
+                          constants.Field.PenaltyShortDist + offset))
+        right_seg = robocup.Segment(
+            robocup.Point(constants.Field.PenaltyLongDist / 2 + offset, 0),
+            robocup.Point(constants.Field.PenaltyLongDist / 2 + offset,
+                          constants.Field.PenaltyShortDist + offset))
+        top_seg = robocup.Segment(
+            robocup.Point(-constants.Field.PenaltyLongDist / 2,
+                          constants.Field.PenaltyShortDist + offset),
+            robocup.Point(constants.Field.PenaltyLongDist / 2,
+                          constants.Field.PenaltyShortDist + offset))
+
+        #Sets the position to mark as the given mark position 
+        #or robot position if no mark position is given
+        pos = self.robot.pos
+        mark_pos = self.mark_point if self.mark_point is not None else self.mark_robot.pos
+
+        #Find best shot point from threat
+        self.kick_eval.add_excluded_robot(self.robot) #FIX: Should we really exclude all of our robots but the goalie?
+        shot_pt, shot_score = self.kick_eval.eval_pt_to_our_goal(mark_pos)
+        self.kick_eval.excluded_robots.clear()
```

Comment:
I think these are just copy paste from the caller function?

Reply:
It's not bad by itself, but if a big chunk of code is being copy pasted and the same calculations are being done, it's a good idea to find a way to merge them. That way if someone wants to make a tweak later on, they don't have to try and edit two places. I'll have to look at the code a little more to see if that's the case here, but it felt a little deja vu looking at the two functions